feat: should make handling transactions

artemklishch opened the pull request:
BREAKING CHANGES:

*Made FruitTransaction class for reading, calculating and writing report regardning transactions *Made methods for calculations depending of trasaction type with strategy pattern *Made Fruit model to grpup information about fruit from transaction *Made Storage for storing calucated data
*Made functions invokes in Main class

Task: https://github.com/mate-academy/jv-fruit-shop

MentorMA commented on src/main/java/core/basesyntax/Main.java:
```diff
@@ -0,0 +1,34 @@
+    private static final String FILE_TO_WRITE_PATH = "src/main/resources/report.csv";
+
+    public static void main(String[] args) {
+        Map<String, FruitHandler> fruitHandlerMap = new HashMap<>();
+        fruitHandlerMap.put(FruitTransaction.Operation.BALANCE.getCode(), new FruitBalance());
+        fruitHandlerMap.put(FruitTransaction.Operation.PURCHASE.getCode(), new FruitPurchase());
+        fruitHandlerMap.put(FruitTransaction.Operation.SUPPLY.getCode(), new FruitSupply());
+        fruitHandlerMap.put(FruitTransaction.Operation.RETURN.getCode(), new FruitReturn());
```
Use map.of

artemklishch replied:
++. changed, thank you

MentorMA commented on src/main/java/core/basesyntax/FruitTransaction.java:
```diff
@@ -0,0 +1,71 @@
+    @Override
+    public void writeReportToFile() {
+        List<String> stringsToReport = new ArrayList<>();
+        stringsToReport.add("fruit,quantity");
```
Save string to constant

artemklishch replied:
++, changed

MentorMA commented on src/main/java/core/basesyntax/FruitTransaction.java:
```diff
@@ -0,0 +1,71 @@
+    }
+
+    @Override
+    public void writeReportToFile() {
+        List<String> stringsToReport = new ArrayList<>();
+        stringsToReport.add("fruit,quantity");
+        for (Fruit fruit : Storage.getFruits().values()) {
+            String string = fruit.getFruit() + "," + fruit.getQuantity();
+            stringsToReport.add(string);
+        }
+        writerService.writeToFile(stringsToReport, fileToWritePath);
+    }
```
In write method you must only write prepared report

artemklishch replied:
++, changed

MentorMA commented on src/main/java/core/basesyntax/FruitTransaction.java:
```diff
@@ -0,0 +1,71 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class FruitTransaction implements FruitTransactionService {
```
Naming of class should be service

artemklishch replied:
++, changed

MentorMA commented on src/main/java/core/basesyntax/FruitTransaction.java:
```diff
@@ -0,0 +1,71 @@
+import java.util.List;
+
+public class FruitTransaction implements FruitTransactionService {
+    private static final ReaderService readerService = new ReaderServiceImpl();
+    private static final WriterService writerService = new WriterServiceImpl();
```
It should be inside methods you don`t need this fields

artemklishch replied:
++, changed

MentorMA commented on src/main/java/core/basesyntax/FruitTransaction.java:
```diff
@@ -0,0 +1,71 @@
+                .toList();
+        for (String[] transactionParts : transactionsOnly) {
+            fruitsTransactionStrategy
+                    .fruitHandler(transactionParts[0])
```
Magic number

artemklishch replied:
++, changed

MentorMA commented on src/main/java/core/basesyntax/FruitTransaction.java:
```diff
@@ -0,0 +1,71 @@
+public class FruitTransaction implements FruitTransactionService {
+    private static final ReaderService readerService = new ReaderServiceImpl();
+    private static final WriterService writerService = new WriterServiceImpl();
+    private final FruitsTransactionStrategy fruitsTransactionStrategy;
+    private final String fileToReadPath;
+    private final String fileToWritePath;
```
Use it like parameters in methods, you don`t need this in fields

artemklishch replied:
++. changed, thanks

MentorMA commented on src/main/java/core/basesyntax/service/impl/WriterServiceImpl.java:
```diff
@@ -0,0 +1,24 @@
+public class WriterServiceImpl implements WriterService {
+    @Override
+    public void writeToFile(List<String> strings, String filePath) {
+        StringBuilder stringBuilder = new StringBuilder();
+        for (String string : strings) {
+            stringBuilder.append(string).append(System.lineSeparator());
+        }
+        File file = new File(filePath);
+        try (BufferedWriter bufferedWriter = new BufferedWriter(new FileWriter(file))) {
+            bufferedWriter.write(stringBuilder.toString());
+        } catch (IOException e) {
```
Write already prepared report

artemklishch replied:
++, changed

MentorMA commented on src/main/java/core/basesyntax/strategy/FruitHandler.java:
```diff
@@ -0,0 +1,8 @@
+package core.basesyntax.strategy;
+
+public interface FruitHandler {
+    int FRUIT_NAME_INDEX = 1;
+    int FRUIT_QUANTITY_INDEX = 2;
```
Is it constant?

artemklishch replied:
++, moved them to another place

MentorMA commented on src/main/java/core/basesyntax/strategy/FruitPurchase.java:
```diff
@@ -0,0 +1,23 @@
+public class FruitPurchase implements FruitHandler {
+    @Override
+    public void transactionHandler(String[] transactionValues) {
+        Map<String, Fruit> fruitsData = Storage.getFruits();
+        String fruitName = transactionValues[FRUIT_NAME_INDEX];
+        int fruitQuantity = Integer.parseInt(transactionValues[FRUIT_QUANTITY_INDEX]);
+        Fruit fruit = fruitsData.get(fruitName);
+        if (fruit == null) {
+            throw new RuntimeException("No fruits in storage to handle purchase");
+        } else if (fruit.getQuantity() < fruitQuantity) {
+            throw new RuntimeException(
+                    "Impossible to handle purchase, because the of the lack of fruits"
+            );
+        }
+        fruit.setQuantity(fruit.getQuantity() - fruitQuantity);
```
In every handler I see duplication of code

artemklishch replied:
++ changed, thank you

artemklishch commented:
Thank you! Updated!

вс, 14 апр. 2024 г. в 13:28, Oleksandr Kapusniak ***@***.***>:

> ***@***.**** requested changes on this pull request.
> ------------------------------
>
> In src/main/java/core/basesyntax/Main.java
> <https://github.com/mate-academy/jv-fruit-shop/pull/1125#discussion_r1564610256>
> :
>
> > +        Map<String, FruitHandler> fruitHandlerMap = new HashMap<>();
> +        fruitHandlerMap.put(FruitTransaction.Operation.BALANCE.getCode(), new FruitBalance());
> +        fruitHandlerMap.put(FruitTransaction.Operation.PURCHASE.getCode(), new FruitPurchase());
> +        fruitHandlerMap.put(FruitTransaction.Operation.SUPPLY.getCode(), new FruitSupply());
> +        fruitHandlerMap.put(FruitTransaction.Operation.RETURN.getCode(), new FruitReturn());
>
> Use map.of
> ------------------------------
>
> In src/main/java/core/basesyntax/FruitTransaction.java
> <https://github.com/mate-academy/jv-fruit-shop/pull/1125#discussion_r1564610490>
> :
>
> > +        List<String> transactionsData = readerService.readFromFile(fileToReadPath);
> +        List<String[]> transactionsOnly = transactionsData.subList(1, transactionsData.size())
> +                .stream()
> +                .map(s -> s.split(","))
> +                .toList();
> +        for (String[] transactionParts : transactionsOnly) {
> +            fruitsTransactionStrategy
> +                    .fruitHandler(transactionParts[0])
> +                    .transactionHandler(transactionParts);
> +        }
> +    }
> +
> +    @Override
> +    public void writeReportToFile() {
> +        List<String> stringsToReport = new ArrayList<>();
> +        stringsToReport.add("fruit,quantity");
>
> Save string to constant
> ------------------------------
>
> In src/main/java/core/basesyntax/FruitTransaction.java
> <https://github.com/mate-academy/jv-fruit-shop/pull/1125#discussion_r1564611057>
> :
>
> > +    public void writeReportToFile() {
> +        List<String> stringsToReport = new ArrayList<>();
> +        stringsToReport.add("fruit,quantity");
> +        for (Fruit fruit : Storage.getFruits().values()) {
> +            String string = fruit.getFruit() + "," + fruit.getQuantity();
> +            stringsToReport.add(string);
> +        }
> +        writerService.writeToFile(stringsToReport, fileToWritePath);
> +    }
>
> In write method you must only write prepared report
> ------------------------------
>
> In src/main/java/core/basesyntax/FruitTransaction.java
> <https://github.com/mate-academy/jv-fruit-shop/pull/1125#discussion_r1564611536>
> :
>
> > @@ -0,0 +1,71 @@
> +package core.basesyntax;
> +
> +import core.basesyntax.db.Storage;
> +import core.basesyntax.model.Fruit;
> +import core.basesyntax.service.FruitTransactionService;
> +import core.basesyntax.service.ReaderService;
> +import core.basesyntax.service.WriterService;
> +import core.basesyntax.service.impl.ReaderServiceImpl;
> +import core.basesyntax.service.impl.WriterServiceImpl;
> +import core.basesyntax.strategy.FruitsTransactionStrategy;
> +import java.util.ArrayList;
> +import java.util.List;
> +
> +public class FruitTransaction implements FruitTransactionService {
>
> Naming of class should be service
> ------------------------------
>
> In src/main/java/core/basesyntax/FruitTransaction.java
> <https://github.com/mate-academy/jv-fruit-shop/pull/1125#discussion_r1564611741>
> :
>
> > +    private static final ReaderService readerService = new ReaderServiceImpl();
> +    private static final WriterService writerService = new WriterServiceImpl();
>
> It should be inside methods you don`t need this fields
> ------------------------------
>
> In src/main/java/core/basesyntax/FruitTransaction.java
> <https://github.com/mate-academy/jv-fruit-shop/pull/1125#discussion_r1564611921>
> :
>
> > +            String fileToWritePath
> +    ) {
> +        this.fruitsTransactionStrategy = fruitsTransactionStrategy;
> +        this.fileToReadPath = fileToReadPath;
> +        this.fileToWritePath = fileToWritePath;
> +    }
> +
> +    public void handleTransactions() {
> +        List<String> transactionsData = readerService.readFromFile(fileToReadPath);
> +        List<String[]> transactionsOnly = transactionsData.subList(1, transactionsData.size())
> +                .stream()
> +                .map(s -> s.split(","))
> +                .toList();
> +        for (String[] transactionParts : transactionsOnly) {
> +            fruitsTransactionStrategy
> +                    .fruitHandler(transactionParts[0])
>
> Magic number
> ------------------------------
>
> In src/main/java/core/basesyntax/FruitTransaction.java
> <https://github.com/mate-academy/jv-fruit-shop/pull/1125#discussion_r1564612285>
> :
>
> > +    private final FruitsTransactionStrategy fruitsTransactionStrategy;
> +    private final String fileToReadPath;
> +    private final String fileToWritePath;
>
> Use it like parameters in methods, you don`t need this in fields
> ------------------------------
>
> In src/main/java/core/basesyntax/service/impl/WriterServiceImpl.java
> <https://github.com/mate-academy/jv-fruit-shop/pull/1125#discussion_r1564612938>
> :
>
> > +        StringBuilder stringBuilder = new StringBuilder();
> +        for (String string : strings) {
> +            stringBuilder.append(string).append(System.lineSeparator());
> +        }
> +        File file = new File(filePath);
> +        try (BufferedWriter bufferedWriter = new BufferedWriter(new FileWriter(file))) {
> +            bufferedWriter.write(stringBuilder.toString());
> +        } catch (IOException e) {
>
> Write already prepared report
> ------------------------------
>
> In src/main/java/core/basesyntax/strategy/FruitHandler.java
> <https://github.com/mate-academy/jv-fruit-shop/pull/1125#discussion_r1564614045>
> :
>
> > +    int FRUIT_NAME_INDEX = 1;
> +    int FRUIT_QUANTITY_INDEX = 2;
>
> Is it constant?
> ------------------------------
>
> In src/main/java/core/basesyntax/strategy/FruitPurchase.java
> <https://github.com/mate-academy/jv-fruit-shop/pull/1125#discussion_r1564614686>
> :
>
> > +        Map<String, Fruit> fruitsData = Storage.getFruits();
> +        String fruitName = transactionValues[FRUIT_NAME_INDEX];
> +        int fruitQuantity = Integer.parseInt(transactionValues[FRUIT_QUANTITY_INDEX]);
> +        Fruit fruit = fruitsData.get(fruitName);
> +        if (fruit == null) {
> +            throw new RuntimeException("No fruits in storage to handle purchase");
> +        } else if (fruit.getQuantity() < fruitQuantity) {
> +            throw new RuntimeException(
> +                    "Impossible to handle purchase, because the of the lack of fruits"
> +            );
> +        }
> +        fruit.setQuantity(fruit.getQuantity() - fruitQuantity);
>
> In every handler I see duplication of code
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/mate-academy/jv-fruit-shop/pull/1125#pullrequestreview-1999541250>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ANMI2YFG2HTKWMBERQO2UVDY5JK3NAVCNFSM6AAAAABGFSZ5KOVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMYTSOJZGU2DCMRVGA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

Rommelua commented on src/main/java/core/basesyntax/Main.java:
```diff
@@ -0,0 +1,34 @@
+        FruitTransactionServiceImpl fruitTransaction = new FruitTransactionServiceImpl(
+                fruitsTransactionStrategy
+        );
+        fruitTransaction.handleTransactions(FILE_TO_READ_PATH);
+        fruitTransaction.writeReportToFile(FILE_TO_WRITE_PATH);
```
Follow requirements to **main** method. We should have 5 operations here, I see only 2 operations.
![image](https://github.com/mate-academy/jv-fruit-shop/assets/20326858/3b4549ee-94cf-4c37-bb22-7c803a2c3d8d)

artemklishch replied:
++, thanks. Updated.

artemklishch commented:
Thank you! Updated!

ср, 17 апр. 2024 г. в 13:18, Leonid Sivko ***@***.***>:

> ***@***.**** requested changes on this pull request.
> ------------------------------
>
> In src/main/java/core/basesyntax/Main.java
> <https://github.com/mate-academy/jv-fruit-shop/pull/1125#discussion_r1568608945>
> :
>
> > +        fruitTransaction.handleTransactions(FILE_TO_READ_PATH);
> +        fruitTransaction.writeReportToFile(FILE_TO_WRITE_PATH);
>
> Follow requirements to *main* method. We should have 5 operations here, I
> see only 2 operations.
> image.png (view on web)
> <https://github.com/mate-academy/jv-fruit-shop/assets/20326858/3b4549ee-94cf-4c37-bb22-7c803a2c3d8d>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/mate-academy/jv-fruit-shop/pull/1125#pullrequestreview-2005658383>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ANMI2YDAJSZOMGHAWBVGUFTY5ZD73AVCNFSM6AAAAABGFSZ5KOVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDAMBVGY2TQMZYGM>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>
